From c6bd6726ade4da581cafa0f1a9195c26c7d9f282 Mon Sep 17 00:00:00 2001
From: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
Date: Tue, 22 Dec 2020 14:12:23 +0100
Subject: [PATCH 024/122] pinctrl: meson: meson8b: move GPIO_BSD_EN to the CBUS
 GPIO controller

Configuring the GPIO_BSD_EN pin is done using different registers:
- output high/low is in CBUS PREG_PAD_GPIO0_O[31]
- pull-up/down is in CBUS PAD_PULL_UP_REG2[0]
- output enable is in CBUS PREG_PAD_GPIO0_O[30]
- unknown purpose: CBUS PREG_PAD_GPIO0_O[29] has to be cleared
- unknown purpose: SECBUS2 AO_SECURE_REG0[0] has to be set

This means five out of six bits needed to configure this pin are in the
CBUS region. Additionally one bit has to be configured which is in the
SECBUS2 region (instead of AOBUS region as the other AOBUS pins).

Move the GPIO_BSD_EN pin to the CBUS group and configure the output
high/low, pull-up/down and output enable bits. The two unknown ones will
be hooked up in the future.

Signed-off-by: Martin Blumenstingl <martin.blumenstingl@googlemail.com>
---
 drivers/pinctrl/meson/pinctrl-meson8b.c | 25 +++++++++++++++++--------
 include/dt-bindings/gpio/meson8b-gpio.h |  5 +++--
 2 files changed, 20 insertions(+), 10 deletions(-)

diff --git a/drivers/pinctrl/meson/pinctrl-meson8b.c b/drivers/pinctrl/meson/pinctrl-meson8b.c
index 39bb0c4b21ed..3089cc80db67 100644
--- a/drivers/pinctrl/meson/pinctrl-meson8b.c
+++ b/drivers/pinctrl/meson/pinctrl-meson8b.c
@@ -100,6 +100,13 @@ static const struct pinctrl_pin_desc meson8b_cbus_pins[] = {
 	MESON_PIN(DIF_3_N),
 	MESON_PIN(DIF_4_P),
 	MESON_PIN(DIF_4_N),
+
+	/*
+	 * The following pin is not mentioned in the public datasheet.
+	 * According to this datasheet it can't be used with the GPIO
+	 * interrupt controller.
+	 */
+	MESON_PIN(GPIO_BSD_EN),
 };
 
 static const struct pinctrl_pin_desc meson8b_aobus_pins[] = {
@@ -119,11 +126,10 @@ static const struct pinctrl_pin_desc meson8b_aobus_pins[] = {
 	MESON_PIN(GPIOAO_13),
 
 	/*
-	 * The following 2 pins are not mentionned in the public datasheet
-	 * According to this datasheet, they can't be used with the gpio
-	 * interrupt controller
+	 * The following pin is not mentioned in the public datasheet.
+	 * According to this datasheet it can't be used with the GPIO
+	 * interrupt controller.
 	 */
-	MESON_PIN(GPIO_BSD_EN),
 	MESON_PIN(GPIO_TEST_N),
 };
 
@@ -440,6 +446,8 @@ static struct meson_pmx_group meson8b_cbus_groups[] = {
 	GPIO_GROUP(DIF_4_P),
 	GPIO_GROUP(DIF_4_N),
 
+	GPIO_GROUP(GPIO_BSD_EN),
+
 	/* bank X */
 	GROUP(sd_d0_a,		8,	5),
 	GROUP(sd_d1_a,		8,	4),
@@ -618,7 +626,6 @@ static struct meson_pmx_group meson8b_aobus_groups[] = {
 	GPIO_GROUP(GPIOAO_11),
 	GPIO_GROUP(GPIOAO_12),
 	GPIO_GROUP(GPIOAO_13),
-	GPIO_GROUP(GPIO_BSD_EN),
 	GPIO_GROUP(GPIO_TEST_N),
 
 	/* bank AO */
@@ -678,14 +685,16 @@ static const char * const gpio_periphs_groups[] = {
 
 	"DIF_0_P", "DIF_0_N", "DIF_1_P", "DIF_1_N",
 	"DIF_2_P", "DIF_2_N", "DIF_3_P", "DIF_3_N",
-	"DIF_4_P", "DIF_4_N"
+	"DIF_4_P", "DIF_4_N",
+
+	"GPIO_BSD_EN"
 };
 
 static const char * const gpio_aobus_groups[] = {
 	"GPIOAO_0", "GPIOAO_1", "GPIOAO_2", "GPIOAO_3",
 	"GPIOAO_4", "GPIOAO_5", "GPIOAO_6", "GPIOAO_7",
 	"GPIOAO_8", "GPIOAO_9", "GPIOAO_10", "GPIOAO_11",
-	"GPIOAO_12", "GPIOAO_13", "GPIO_BSD_EN", "GPIO_TEST_N"
+	"GPIOAO_12", "GPIOAO_13", "GPIO_TEST_N"
 };
 
 static const char * const sd_a_groups[] = {
@@ -942,7 +951,7 @@ static struct meson_bank meson8b_cbus_banks[] = {
 	 * - clears CBUS PREG_PAD_GPIO0_O[29]
 	 * - sets SECBUS2 AO_SECURE_REG0[0]
 	 */
-	BANK("BSD_EN",	 GPIO_BSD_EN,	GPIO_BSD_EN,-1, -1,  0, -1,   2, 31,   1, 30,  1, 31,    0, -1),
+	BANK("BSD_EN",	 GPIO_BSD_EN,	GPIO_BSD_EN, -1, -1,  0, -1,   2, 31,   1, 30,   1, 31,   0, -1),
 };
 
 static struct meson_bank meson8b_aobus_banks[] = {
diff --git a/include/dt-bindings/gpio/meson8b-gpio.h b/include/dt-bindings/gpio/meson8b-gpio.h
index 7c3bc0782eca..ea8a5502e3e0 100644
--- a/include/dt-bindings/gpio/meson8b-gpio.h
+++ b/include/dt-bindings/gpio/meson8b-gpio.h
@@ -100,6 +100,8 @@
 #define DIF_4_P		81
 #define DIF_4_N		82
 
+#define GPIO_BSD_EN	83
+
 /* AO GPIO chip */
 #define GPIOAO_0	0
 #define GPIOAO_1	1
@@ -115,7 +117,6 @@
 #define GPIOAO_11	11
 #define GPIOAO_12	12
 #define GPIOAO_13	13
-#define GPIO_BSD_EN	14
-#define GPIO_TEST_N	15
+#define GPIO_TEST_N	14
 
 #endif /* _DT_BINDINGS_MESON8B_GPIO_H */
-- 
2.25.1

